lib/src/df.dart

```diff
@@ -355,82 +351,110 @@ class DataFrame {
+
+  /// In-place sort this dataframe by a column
+  /// TODO(caseycrogers): Add support for custom comparator functions.
+  void sort(String colName, {required NullSortBehavior nullBehavior}) =>
+      _sort(colName, inPlace: true, nullBehavior: nullBehavior);
+
+  List<List<Object?>> _sort(String colName,
+      {required bool inPlace, required NullSortBehavior nullBehavior}) {
+    final newMatrixData = inPlace
+        ? _matrix.data
+        // Deep copy the rows.
+        : _matrix.data.map((row) => List<Object?>.from(row)).toList();
+    return newMatrixData
+      ..sort((a, b) => _compareRows(a, b, columnIndex(colName), nullBehavior));
+  }
 
-  dynamic _sort(String colName, {@required bool inPlace}) {
-    assert(colName != null);
-    final colData =
-        _matrix.typedRecordsForColumnIndice<dynamic>(_indiceForColumn(colName));
-    // create a map of index/data
-    final dataIndex = <int, dynamic>{};
-    var i = 0;
-    colData.forEach((dynamic record) {
-      dataIndex[i] = record;
-      ++i;
-    });
-    // sort the index map from values
-    final sortedKeys = dataIndex.keys.toList(growable: false)
-      ..sort((k1, k2) => (dataIndex[k1] as dynamic)
-          .compareTo(dataIndex[k2] as dynamic) as int);
-    final sortedMap = LinkedHashMap<int, dynamic>.fromIterable(sortedKeys,
-        key: (dynamic k) => k as int, value: (dynamic k) => dataIndex[k]);
-    final order = sortedMap.keys;
-    // rebuild the dataset in order
-    final _newMatrix = <List<dynamic>>[];
-    for (final i in order) {
-      _newMatrix.add(_matrix.data[i]);
-    }
-    if (!inPlace) {
-      return DataFrame._copyWithMatrix(this, _newMatrix);
-    } else {
-      _matrix.data = _newMatrix;
-    }
-    return null;
+  int _compareRows(List<Object?> rowA, List<Object?> rowB, int index,
+      NullSortBehavior nullBehavior) {
+    final recordA = _matrix.typedRecordForColumnIndexInRow<Comparable<Object>?>(
+            index, rowA) ??
+        _replaceNullForSort(nullBehavior);
+    final recordB = _matrix.typedRecordForColumnIndexInRow<Comparable<Object>?>(
+            index, rowA) ??
+        _replaceNullForSort(nullBehavior);
+    return recordA.compareTo(recordB);
   }
 
-  int _indiceForColumn(String colName) {
-    int ind;
+  int _indexForColumn(String colName) {
+    int? ind;
     var i = 0;
     for (final col in _columns) {
       if (colName == col.name) {
         ind = i;
         break;
       }
-      ++i;
+      i++;
     }
     if (ind == null) {
       throw ColumnNotFoundException('Can not find column $colName');
     }
     return ind;
   }
 
-  Map<int, String> _columnsIndices() {
-    final ind = <int, String>{};
-    var i = 0;
-    for (final col in _columns) {
-      ind[i] = col.name;
-      ++i;
-    }
-    return ind;
+  List<String> _columnIndices() => columns.map((c) => c.name).toList();
+}
+
+/// How to treat nulls when taking the average over a column.
+enum NullMeanBehavior {
```

Comment:
I've added arguments to `sort` and `mean` to require users to explicitly specify how nulls should be handled when aggregated a column.

Reply:
I guess we will have to document these changes when released